File: src/main/java/net/hellomouse/progression_change/ProgressionModConfig.java
package net.hellomouse.progression_change;

import me.shedaniel.clothconfig2.api.ConfigBuilder;
import me.shedaniel.clothconfig2.api.ConfigCategory;
import me.shedaniel.clothconfig2.api.ConfigEntryBuilder;
import net.minecraft.text.Text;

public class ProgressionModConfig {

    public static OreDropChanges oreDropChanges = new OreDropChanges();
    public static EarlyGameChanges earlyGameChanges = new EarlyGameChanges();

    public static class OreDropChanges {
        public boolean moddedPickaxeWorkaround = true; // if a mod didn't registerTier to TierSortingRegistry, check the mining level int instead
        public int rawCopperNuggetDrop = 1;
        public int rawIronNuggetDrop = 1;
        public int rawGoldNuggetDrop = 1;
        public int diamondFragmentDrop = 1;
        public boolean oreToStone = false;
    }

    public static class EarlyGameChanges {
        public boolean overridePlantFiberProbability = false;
        public int plantFiberDropProbability = 5;
        public boolean overridePebbleDropProbability = false;
        public int pebbleDropProbability = 40;
    }

    public static class Gui {
        private static Text getTranslatableText(String name) {
            return Text.translatable("text.config." + ProgressionMod.MODID + "." + name);
        }

        private static Text getTranslatableTextOption(String name) {
            return getTranslatableText("option." + name);
        }

        public static ConfigBuilder getConfigBuilder() {
            ConfigBuilder builder = ConfigBuilder.create().setTitle(getTranslatableText("title")).transparentBackground();
            ConfigEntryBuilder entryBuilder = ConfigEntryBuilder.create();
            addOreDropChangesEntries(builder, entryBuilder);
            addEarlyGameEntries(builder, entryBuilder);
            return builder; // heh, rewriting the entire config just for this
        }

        private static void addEarlyGameEntries(ConfigBuilder configBuilder, ConfigEntryBuilder entryBuilder) {
            ConfigCategory category = configBuilder.getOrCreateCategory(getTranslatableTextOption("earlyGameChanges"));
            category.addEntry(
                    entryBuilder.startBooleanToggle(
                                    getTranslatableTextOption("earlyGameChanges.overridePlantFiberDropProbability"),
                                    earlyGameChanges.overridePlantFiberProbability
                            ).setSaveConsumer(aBoolean -> earlyGameChanges.overridePlantFiberProbability = aBoolean)
                            .setDefaultValue(false)
                            .build()
            );
            category.addEntry(
                    entryBuilder.startIntSlider(
                                    getTranslatableTextOption("earlyGameChanges.plantFiberDropProbability"),
                                    earlyGameChanges.plantFiberDropProbability,
                                    1,
                                    100
                            ).setSaveConsumer(aInt -> earlyGameChanges.plantFiberDropProbability = aInt)
                            .setDefaultValue(5)
                            .build()
            );
            category.addEntry(
                    entryBuilder.startBooleanToggle(
                                    getTranslatableTextOption("earlyGameChanges.overridePebbleDropProbability"),
                                    earlyGameChanges.overridePebbleDropProbability
                            ).setSaveConsumer(aBoolean -> earlyGameChanges.overridePebbleDropProbability = aBoolean)
                            .setDefaultValue(false)
                            .build()
            );
            category.addEntry(
                    entryBuilder.startIntSlider(
                                    getTranslatableTextOption("earlyGameChanges.pebbleDropProbability"),
                                    earlyGameChanges.pebbleDropProbability,
                                    1,
                                    100
                            ).setSaveConsumer(aInt -> earlyGameChanges.pebbleDropProbability = aInt)
                            .setDefaultValue(40)
                            .build()
            );
        }

        private static void addOreDropChangesEntries(ConfigBuilder configBuilder, ConfigEntryBuilder entryBuilder) {
            ConfigCategory category = configBuilder.getOrCreateCategory(getTranslatableTextOption("oreDropChanges"));
            category.addEntry(
                    entryBuilder.startBooleanToggle(
                                    getTranslatableTextOption("oreDropChanges.moddedPickaxeWorkaround"), oreDropChanges.moddedPickaxeWorkaround)
                            .setSaveConsumer(aBoolean -> oreDropChanges.moddedPickaxeWorkaround = aBoolean)
                            .setTooltip(getTranslatableTextOption("oreDropChanges.moddedPickaxeWorkaround.tooltip"))
                            .setDefaultValue(true)
                            .build());
            category.addEntry(
                    entryBuilder.startIntSlider(
                                    getTranslatableTextOption("oreDropChanges.rawCopperNuggetDrop"), oreDropChanges.rawCopperNuggetDrop, 1, 9)
                            .setSaveConsumer(aInt -> oreDropChanges.rawCopperNuggetDrop = aInt)
                            .setDefaultValue(1)
                            .build());
            category.addEntry(
                    entryBuilder.startIntSlider(
                                    getTranslatableTextOption("oreDropChanges.rawIronNuggetDrop"), oreDropChanges.rawIronNuggetDrop, 1, 9)
                            .setSaveConsumer(aInt -> oreDropChanges.rawIronNuggetDrop = aInt)
                            .setDefaultValue(1)
                            .build());
            category.addEntry(
                    entryBuilder.startIntSlider(
                                    getTranslatableTextOption("oreDropChanges.rawGoldNuggetDrop"), oreDropChanges.rawGoldNuggetDrop, 1, 9)
                            .setSaveConsumer(aInt -> oreDropChanges.rawGoldNuggetDrop = aInt)
                            .setDefaultValue(1)
                            .build());
            category.addEntry(
                    entryBuilder.startIntSlider(
                                    getTranslatableTextOption("oreDropChanges.rawDiamondFragmentDrop"), oreDropChanges.diamondFragmentDrop, 1, 9)
                            .setSaveConsumer(aInt -> oreDropChanges.diamondFragmentDrop = aInt)
                            .setDefaultValue(1)
                            .build());
            category.addEntry(
                    entryBuilder.startBooleanToggle(
                                    getTranslatableTextOption("oreDropChanges.oreToStone"), oreDropChanges.oreToStone)
                            .setSaveConsumer(aBoolean -> oreDropChanges.oreToStone = aBoolean)
                            .setTooltip(getTranslatableTextOption("oreDropChanges.oreToStone.tooltip"))
                            .setDefaultValue(false)
                            .build());
        }
    }
}

File: src/main/kotlin/net/hellomouse/xeno_early_start/ProgressionModConfig.kt
package net.hellomouse.xeno_early_start

object ProgressionModConfig {
    class Config {
        @JvmField
        var oreDropChanges: OreDropChanges = OreDropChanges()

        @JvmField
        var earlyGameChanges: EarlyGameChanges = EarlyGameChanges()

        @JvmField
        var mobChanges: MobChanges = MobChanges()

        @JvmField
        var blockChanges: BlockChanges = BlockChanges()
    }

    @JvmField
    var config: Config = Config()

    class OreDropChanges {
        var rawCopperNuggetDrop: Int = 1

        var rawIronNuggetDrop: Int = 1

        var rawGoldNuggetDrop: Int = 1

        var diamondFragmentDrop: Int = 1

        var oreToStone: Boolean = false
    }

    class EarlyGameChanges {
        var overridePlantFiberProbability: Boolean = false

        var plantFiberDropProbability: Int = 5

        var overridePebbleDropProbability: Boolean = false

        var pebbleDropProbability: Int = 40
    }

    class MobChanges {
        var flatAdditiveMobSpawnWithEquipment: Float = 0.05f

        var replaceEntityCopperArmourProbability: Float = 0.05f

        var entitySpawnWithCopperToolProbability: Float = 0.05f

        var polarBearAlwaysAggressive: Boolean = true

        var polarBearSpeed: Float = 0.35f

        var polarBearRange: Int = 40

        var mobAttackWeakPlayer: Boolean = true

        var wolfAggressiveAtNight: Boolean = true

    }

    class BlockChanges {
        enum class FixThinBlockStepSound {
            True,
            False,
            OnlyThisMod
        }

        var stonecutterDamage: Float = 3f
        var coralDamage: Float = 1.5f
        var amethystFallDamageMultiplier: Float = 1.5f
        var cookerDamage: Float = 0.5f
        var brickFurnaceDamage: Float = 1f
        var furnaceDamage: Float = 1.5f
        var blastFurnaceDamage: Float = 2.5f
        var fixThinBlockStepSound: FixThinBlockStepSound = FixThinBlockStepSound.True
    }

}
